src/server.rs

```diff
@@ -2334,116 +2339,103 @@ errorCounts
         }
     }
 
-    #[test]
-    async fn test_hover() {
-        let fluxscript = r#"x = 1
-x + 1
-"#;
+    #[track_caller]
+    async fn test_hover(fluxscript: &str, expected: &str) {
```

Comment:
I'd argue the boilerplate is irrelevant in this case, what we are testing for is that l the location under the cursor returns the relevant type, the boiler plate to get there isn't very interesting. That is precisely described when looking at a call to  `test_hover`.

Reply:
My take is the narrative aspect is improved with this change since each case now reads closer to how I might verbally describe the interaction.

I grant that conventional wisdom says "let tests be WET" so they are more isolated and self evident, but in this case I think an assertion helper is welcome.

@rockstar would it be easier to accept as a separate piece of code elsewhere, advertised as a test helper? In JS UI testing, it's common place to have libraries of helpers to better support phrasing test assertions in terms of user interactions. This feels similar to me.

Reply:
I disagree. The previous test style, I could see the input, the interaction, and the expected output without pushing anything onto my mental stack. For instance, this test...

        async fn test_hover_simple() {
            test_hover(
                r#"xyzw = 1
    xyzw + 1
    // ^
    "#,
                "type: int",
            )
            .await;
        }

...pushes another thing onto the mental stack. This _seems_ shallow, until all your tests do multiple levels of this indirection. I don't think helpers are good _for the actual test_. Test setup/teardown, sure. Maybe it's even okay in the assertions (though that gets hairy _really_ quick). I don't feel it's wise to reduce the test itself into a single call to a helper.

Reply:
Separated to https://github.com/influxdata/flux-lsp/pull/386 .

I don't think helpers add complexity here. If you know how the helper works, you can just look at the call and immediately see what the test is about. If you don't know how it works then you might not need to or you just jump into the helper to check what it does and it is the same amount of code anyway. Once you do know what it does, skimming the other tests is easier since you know they have the same setup/teardown. If it is duplicated you tend to have to check that they don't do things slightly differently.

It also makes the tests less vulnerable to refactors as the duplication would force more changes.

Reply:
> If you know how the helper works

This is the point I'm trying to make: when looking at a test, I would _prefer_ that we can see exactly what needs to happen in the test and not have to refer to another function. This seems like a really small thing, but I think the flux test suite is a prime example of what I'm trying to avoid: you often have to push onto your mental stack three or four times before you can really get an idea what the test is actually trying to do.